### ISSUE TITLE
The Gradle wrapper shell script is of low quality

### ISSUE THREAD

**Issue by rdiez** [gradle/gradle#1897]
The Gradle wrapper shell script is of low quality:
- It is brittle, because it does not stop on error. It just blindly carries on.
- It does not error on undefined variables, which makes it harder to maintain and modify.
- It does not output warnings and errors to stderr instead of stdout.
- It only handles up to 9 command-line arguments on Cygwin (for wrapper version 2.14).
- It does not mention that it has been automatically generated. If you open a third-party Java project and don't know much about Gradle, you may be tempted to modify it by hand. This has actually happened to me!

In Gradle version 3.xx, the wrapper has switched from bash to a generic POSIX shell, which makes it even harder to get it right, especially since "set -o pipefail" is not available then.

This is an example of an improved wrapper for Gradle version 2.14:

```
#!/usr/bin/env bash

##############################################################################
##
##  Gradle start up script for UN*X
##
##############################################################################

set -o errexit
set -o nounset
set -o pipefail


# Add default JVM options here. You can also use JAVA_OPTS and GRADLE_OPTS to pass JVM options to this script.
DEFAULT_JVM_OPTS=""

APP_NAME="Gradle"
APP_BASE_NAME="$(basename "$BASH_SOURCE")"

# Use the maximum available, or set MAX_FD != -1 to use that value.
MAX_FD="maximum"

warn ( ) {
    echo "$*" >&2
}

die ( ) {
    echo >&2
    echo "$*" >&2
    echo >&2
    exit 1
}

is_var_set_and_not_empty ()
{
  if [ -z "${!1:-}" ]; then return 1; else return 0; fi
}

# OS specific support (must be 'true' or 'false').
cygwin=false
msys=false
darwin=false
case "`uname`" in
  CYGWIN* )
    cygwin=true
    ;;
  Darwin* )
    darwin=true
    ;;
  MINGW* )
    msys=true
    ;;
esac

# For Cygwin, ensure paths are in UNIX format before anything is touched.
if $cygwin ; then
    is_var_set_and_not_empty JAVA_HOME && JAVA_HOME=`cygpath --unix "$JAVA_HOME"`
fi

# Attempt to set APP_HOME
# Resolve links: $0 may be a link
PRG="$0"
# Need this for relative symlinks.
while [ -h "$PRG" ] ; do
    ls=`ls -ld "$PRG"`
    link=`expr "$ls" : '.*-> \(.*\)$'`
    if expr "$link" : '/.*' > /dev/null; then
        PRG="$link"
    else
        PRG=`dirname "$PRG"`"/$link"
    fi
done
SAVED="`pwd`"
cd "`dirname \"$PRG\"`/" >&-
APP_HOME="`pwd -P`"
cd "$SAVED" >&-

CLASSPATH="$APP_HOME/gradle/wrapper/gradle-wrapper.jar"

# Determine the Java command to use to start the JVM.
if is_var_set_and_not_empty JAVA_HOME ; then
    if [ -x "$JAVA_HOME/jre/sh/java" ] ; then
        # IBM's JDK on AIX uses strange locations for the executables
        JAVACMD="$JAVA_HOME/jre/sh/java"
    else
        JAVACMD="$JAVA_HOME/bin/java"
    fi
    if [ ! -x "$JAVACMD" ] ; then
        die "ERROR: JAVA_HOME is set to an invalid directory: $JAVA_HOME

Please set the JAVA_HOME variable in your environment to match the
location of your Java installation."
    fi
else
    JAVACMD="java"
    which "$JAVACMD" >/dev/null 2>&1 || die "ERROR: JAVA_HOME is not set and no '$JAVACMD' command could be found in your PATH.

Please set the JAVA_HOME variable in your environment to match the
location of your Java installation."
fi

# Increase the maximum file descriptors if we can.
if [ "$cygwin" = "false" -a "$darwin" = "false" ] ; then
    MAX_FD_LIMIT=`ulimit -H -n`
    if [ $? -eq 0 ] ; then
        if [ "$MAX_FD" = "maximum" -o "$MAX_FD" = "max" ] ; then
            MAX_FD="$MAX_FD_LIMIT"
        fi
        ulimit -n $MAX_FD
        if [ $? -ne 0 ] ; then
            warn "Could not set maximum file descriptor limit: $MAX_FD"
        fi
    else
        warn "Could not query maximum file descriptor limit: $MAX_FD_LIMIT"
    fi
fi

# For Darwin, add options to specify how the application appears in the dock
if $darwin; then
    GRADLE_OPTS="${GRADLE_OPTS:-} \"-Xdock:name=$APP_NAME\" \"-Xdock:icon=$APP_HOME/media/gradle.icns\""
fi

# For Cygwin, switch paths to Windows format before running java
if $cygwin ; then
    APP_HOME=`cygpath --path --mixed "$APP_HOME"`
    CLASSPATH=`cygpath --path --mixed "$CLASSPATH"`

    # We build the pattern for arguments to be converted via cygpath
    ROOTDIRSRAW=`find -L / -maxdepth 1 -mindepth 1 -type d 2>/dev/null`
    SEP=""
    for dir in $ROOTDIRSRAW ; do
        ROOTDIRS="${ROOTDIRS:-}$SEP$dir"
        SEP="|"
    done
    # This pattern matches strings that start with an absolute path like "/usr".
    # Paths embedded in an argument will be missed. For example, an argument like "--prefix=/usr" will not be converted to a Windows path.
    OURCYGPATTERN="(^($ROOTDIRS))"
    # Add a user-defined pattern to the cygpath arguments
    if is_var_set_and_not_empty GRADLE_CYGPATTERN; then
        OURCYGPATTERN="$OURCYGPATTERN|($GRADLE_CYGPATTERN)"
    fi

    declare -a CONVERTED_ARGS
    for arg in "$@" ; do
      if [[ $arg =~ $OURCYGPATTERN ]]; then
        CONVERTED_ARGS+=("$(cygpath --path --ignore --mixed "$arg")")
      else
        CONVERTED_ARGS+=("$arg")
      fi
    done

    set -- "${CONVERTED_ARGS[@]}"

    if false; then
      printf -v QUOTED_ARGS "%q  " "$@"
      echo "Found $# arguments: $QUOTED_ARGS"
    fi
fi

# Split up the JVM_OPTS And GRADLE_OPTS values into an array, following the shell quoting and substitution rules
function splitJvmOpts() {
    JVM_OPTS=("$@")
}
eval splitJvmOpts $DEFAULT_JVM_OPTS ${JAVA_OPTS:-} ${GRADLE_OPTS:-}
JVM_OPTS[${#JVM_OPTS[*]}]="-Dorg.gradle.appname=$APP_BASE_NAME"

exec "$JAVACMD" "${JVM_OPTS[@]}" -classpath "$CLASSPATH" org.gradle.wrapper.GradleWrapperMain "$@"
```

**Comment by lacasseio** [comment]
Thanks for the issue @rdiez. Would you care to contribute a  PR for this issue. See the [contributing guidelines](https://github.com/gradle/gradle/blob/master/.github/CONTRIBUTING.md) for more information.

**Comment by rdiez** [comment]
I am sorry, but I really do not have the time to read through your contributors agreement etc. and find where your generator is located. Most changes are pretty trivial anyway:

1) Switch back to bash in the shebang, so that you can use the following like I showed you in the script above:

set -o errexit
set -o nounset
set -o pipefail

With the POSIX shell there is no easy way to stop when a process in a pipeline fails.

2) Add ">&2" to your "warn" and "die" functions, see above.

3) If you care about Cygwin, diff my script above against your 2.14 version to see the changes in this area. The diff will also show other tiny changes you may want to incorporate.

4) Just add a comment to mention that the script has been automatically generated and should not be modified by hand.

You don't have to do all changes above, any of them would be an improvement.

**Comment by gimoh** [comment]
Just in case the wrapper gets some love at some point ;) it would be worth looking into not using `which` as well as it is not available in all environments (e.g. minimal docker images). Generally AFAIK you can use `type java` to check if java is in PATH and that should be more available as it's a bourne shell builtin.

**Comment by texasbruce** [comment]
Additionally it checks for MINGW* and set msys=true, but never test and use the variable, and it is broken under MINGW/MSYS2

Was the PR submitted yet?

**Comment by yukihane** [comment]
It seems MINGW* ploblem is fixed on #8679 .
